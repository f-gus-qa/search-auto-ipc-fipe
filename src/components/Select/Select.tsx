import { isEqual } from 'lodash';
import { Autocomplete } from '@material-ui/lab';
import { CircularProgress, TextField } from '@material-ui/core';

interface SelectProps<T> {
  options: Array<T>,
  visualOptionName: keyof T,
  inputLabel: string,
  selected: T | null,
  disabled?: boolean,
  isLoading?: boolean,
  onChange: (arg0: T | null) => void
}

function Select<T>({
  options,
  visualOptionName,
  inputLabel,
  disabled,
  selected,
  onChange,
  isLoading,
}: SelectProps<T>) {
  return (
    <Autocomplete
      value={selected}
      options={options}
      disabled={disabled}
      onChange={(ev, newValue) => onChange(newValue)}
      style={{ maxWidth: 450, width: '100%' }}
      getOptionLabel={(option) => String(option[visualOptionName])}
      getOptionSelected={(option, value) => isEqual(option, value)}
      renderInput={(params) => (
        <TextField
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...params}
          label={inputLabel}
          variant="outlined"
          InputProps={{
            ...params.InputProps,
            endAdornment: (
              <>
                {isLoading ? <CircularProgress data-testid="progressLoadingInput" color="inherit" size={20} /> : null}
                {params.InputProps.endAdornment}
              </>
            ),
          }}
        />
      )}
    />
  );
}

Select.defaultProps = {
  disabled: false,
  isLoading: false,
};

export default Select;
