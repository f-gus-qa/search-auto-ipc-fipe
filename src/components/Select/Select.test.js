/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { Autocomplete } from '@material-ui/lab';
import Select from '.';

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  TextField: jest.fn().mockImplementation(
    (props) => (
      <div>{props.InputProps.endAdornment}</div>
    ),
  ),
}));
jest.mock('@material-ui/lab', () => ({
  Autocomplete: jest.fn().mockImplementation(
    (props) => {
      const ActualAutocomplete = jest.requireActual('@material-ui/lab').Autocomplete;
      return <ActualAutocomplete {...props} />;
    },
  ),
}));

const onChange = jest.fn();

const makeAutocomplete = ({
  disabled,
  isLoading,
} = {
  disabled: false,
  isLoading: false,
}) => (
  <Select
    options={[]}
    inputLabel="some-label"
    onChange={onChange}
    selected={null}
    visualOptionName="some-field"
    disabled={disabled}
    isLoading={isLoading}
  />
);

describe('<Select>', () => {
  it('check renders', () => {
    const { toJSON } = create(makeAutocomplete());

    expect(toJSON()).toMatchSnapshot();
  });

  describe('when onChange has been called', () => {
    beforeEach(() => {
      Autocomplete.mockImplementationOnce(
        (props) => {
          props.onChange('some-event', 'some-option');
          return <></>;
        },
      );
    });

    it('check onChange', () => {
      render(makeAutocomplete());

      expect(onChange).toBeCalledWith('some-option');
    });
  });

  describe('when disable select', () => {
    beforeEach(() => {
      Autocomplete.mockImplementationOnce();
    });

    it('check renders', () => {
      const { toJSON } = create(makeAutocomplete({ disabled: true, isLoading: false }));

      expect(toJSON()).toMatchSnapshot();
    });

    it('check if disabled props has been passed', () => {
      render(makeAutocomplete({ disabled: true, isLoading: false }));

      const { disabled } = Autocomplete.mock.calls[0][0];

      expect(disabled).toBeTruthy();
    });
  });

  describe('when select is loading', () => {
    beforeEach(() => {
      Autocomplete.mockImplementationOnce();
    });

    it('check renders', () => {
      const { toJSON } = create(makeAutocomplete({ disabled: true, isLoading: false }));

      expect(toJSON()).toMatchSnapshot();
    });

    it('check if loading render', () => {
      render(makeAutocomplete({ disabled: false, isLoading: true }));

      expect(screen.getByTestId('progressLoadingInput')).toBeInTheDocument();
    });
  });
});
