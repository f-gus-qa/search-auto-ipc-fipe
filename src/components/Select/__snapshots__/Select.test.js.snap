// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select> check renders 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
  style={
    Object {
      "maxWidth": 450,
      "width": "100%",
    }
  }
>
  <div>
    <div
      className="MuiAutocomplete-endAdornment"
    >
      <button
        aria-label="Clear"
        className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        title="Clear"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
      </button>
      <button
        aria-label="Open"
        className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        title="Open"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Select> when disable select check renders 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
  style={
    Object {
      "maxWidth": 450,
      "width": "100%",
    }
  }
>
  <div>
    <div
      className="MuiAutocomplete-endAdornment"
    >
      <button
        aria-label="Open"
        className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        title="Open"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Select> when select is loading check renders 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
  style={
    Object {
      "maxWidth": 450,
      "width": "100%",
    }
  }
>
  <div>
    <div
      className="MuiAutocomplete-endAdornment"
    >
      <button
        aria-label="Open"
        className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        title="Open"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
